exp/controllers/awsmachinepool_controller.go

```diff
@@ -455,6 +455,11 @@ func (r *AWSMachinePoolReconciler) reconcileLaunchTemplate(machinePoolScope *sco
 	// userdata, OR we've discovered a new AMI ID.
 	if needsUpdate || tagsChanged || *imageID != *launchTemplate.AMI.ID || launchTemplateUserDataHash != bootstrapDataHash {
 		machinePoolScope.Info("creating new version for launch template", "existing", launchTemplate, "incoming", machinePoolScope.AWSMachinePool.Spec.AWSLaunchTemplate)
+		// There is a limit to the number of Launch Template Versions.
+		// We ensure that the number of versions does not grow without bound by following a simple rule: For each version we create, we delete one old version.
+		if err := ec2svc.PruneLaunchTemplateVersions(machinePoolScope.AWSMachinePool.Status.LaunchTemplateID); err != nil {
```

Comment:
Is it better to create one before pruning? In case creation fails we don't want to delete the previous one.
We create a new one, it is tagged as latest, so the previous can be deleted.

`CreateLaunchTemplateVersion` returns the version created, how about directly trying to delete the previous version? Assuming the numbering is strictly increasing.

Reply:
> Is it better to create one before pruning?

If we create one before removing one, then we may end up never removing any. Whether or not we return an error when we fail to remove one, we'll create more than we'll remove, and therefore eventually the limit.

> In case creation fails we don't want to delete the previous one.
> We create a new one, it is tagged as latest, so the previous can be deleted.

I took care to never remove the version that is currently "in use." Please see the `PruneLaunchTemplateVersions` godoc, and let me know if anything is unclear. Thanks!

Reply:
I've revised the comment at this line. I hope it's more clear now.